File: java/com/google/zetasql/resolvedast/ResolvedColumn.java
/*
 * Copyright 2019 ZetaSQL Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.google.zetasql.resolvedast;

import com.google.common.base.Preconditions;
import com.google.zetasql.Type;
import java.io.Serializable;
import java.util.List;
import java.util.Objects;

/**
 * A column produced by part of a query (e.g. a scan or subquery).
 *
 * <p>This is used in the column_list of a resolved AST node to represent a "slot" in the "tuple"
 * produced by that logical operator. This is also used in expressions in ResolvedColumnRef to point
 * at the column selected during name resolution.
 *
 * <p>The column_id is the definitive identifier for a ResolvedColumn, and the column_id should be
 * used to match a ResolvedColumnRef in an expression to the scan that produces that column.
 * column_ids are unique within a query. If the same table is scanned multiple times, distinct
 * column_ids will be chosen for each scan.
 *
 * <p>Joins and other combining nodes may propagate ResolvedColumns from their inputs, with the same
 * column_ids.
 */
public final class ResolvedColumn implements Serializable {

  private final long id;
  private final String tableName;
  private final String name;
  private final Type type;

  public ResolvedColumn(long id, String tableName, String name, Type type) {
    this.id = id;
    this.tableName = Preconditions.checkNotNull(tableName);
    this.name = Preconditions.checkNotNull(name);
    this.type = Preconditions.checkNotNull(type);
  }

  public long getId() {
    return id;
  }

  public String getTableName() {
    return tableName;
  }

  public String getName() {
    return name;
  }

  public Type getType() {
    return type;
  }

  public String debugString() {
    return tableName + "." + name + "#" + id;
  }

  public String shortDebugString() {
    return name + "#" + id;
  }

  @Override
  public String toString() {
    return debugString();
  }

  public boolean isDefaultValue() {
    return id == -1;
  }

  /**
   * Format a list of ResolvedColumns. Uses an abbreviated outputformat if all columns have the same
   * tableName.
   *
   * <p>The implementation is nearly identical to the C++ implementation in
   * ResolvedColumnListToString. Since the outputs must be the same, the implementations should be
   * similar to make changes to the output easier.
   */
  public static String toString(List<ResolvedColumn> columns) {
    if (columns.isEmpty()) {
      return "[]";
    }
    String commonTableName = columns.get(0).tableName;
    // Use the regular format if we have only one column.
    boolean useCommonTableName = columns.size() > 1;
    for (int i = 1; i < columns.size(); i++) {
      if (!columns.get(i).tableName.equals(commonTableName)) {
        useCommonTableName = false;
        break;
      }
    }

    StringBuilder sb = new StringBuilder();
    if (useCommonTableName) {
      sb.append(commonTableName).append(".[");
      for (ResolvedColumn column : columns) {
        if (column != columns.get(0)) {
          sb.append(", ");
        }
        sb.append(column.shortDebugString());
      }
    } else {
      sb.append("[");
      for (ResolvedColumn column : columns) {
        if (column != columns.get(0)) {
          sb.append(", ");
        }
        sb.append(column.debugString());
      }
    }
    sb.append("]");
    return sb.toString();
  }

  @Override
  /** To provide parity with resolved_column.h, equality is defined using columnId only. */
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }

    if (o instanceof ResolvedColumn) {
      ResolvedColumn other = (ResolvedColumn) o;
        return Objects.equals(this.id, other.id);
    }
    return false;
  }

  @Override
  /** To provide parity with resolved_column.h, hashCode is defined using columnId only. */
  public int hashCode() {
    return Objects.hashCode(id);
  }
}

File: zetasql/resolved_ast/resolved_column.cc
//
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include "zetasql/resolved_ast/resolved_column.h"

#include <memory>

#include "zetasql/resolved_ast/serialization.pb.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "zetasql/base/status_macros.h"

namespace zetasql {

std::string ResolvedColumn::DebugString() const {
  return absl::StrCat(
      table_name_.ToStringView(), ".", name_.ToStringView(), "#", column_id_,
      type_annotation_map() == nullptr ? ""
                                       : type_annotation_map()->DebugString());
}

std::string ResolvedColumn::ShortDebugString() const {
  return absl::StrCat(name_.ToStringView(), "#", column_id_);
}

absl::Status ResolvedColumn::SaveTo(
    FileDescriptorSetMap* file_descriptor_set_map,
    ResolvedColumnProto* proto) const {
  // Consider serializing ResolvedColumn in a separate table, indexed by
  // column_id_, and then only serialize the column_id_ in the AST.
  proto->set_table_name(std::string(table_name_.ToStringView()));
  proto->set_name(std::string(name_.ToStringView()));

  proto->set_column_id(column_id_);
  if (type_annotation_map() != nullptr) {
    ZETASQL_RETURN_IF_ERROR(
        type_annotation_map()->Serialize(proto->mutable_annotation_map()));
  }
  return type()->SerializeToProtoAndDistinctFileDescriptors(
      proto->mutable_type(), file_descriptor_set_map);
}

absl::StatusOr<ResolvedColumn> ResolvedColumn::RestoreFrom(
    const ResolvedColumnProto& proto,
    const ResolvedNode::RestoreParams& params) {
  const IdString table_name = params.string_pool->Make(proto.table_name());
  const IdString column_name = params.string_pool->Make(proto.name());
  const Type* type;
  ZETASQL_RETURN_IF_ERROR(params.type_factory->DeserializeFromProtoUsingExistingPools(
      proto.type(), params.pools, &type));
  const AnnotationMap* annotation_map = nullptr;
  if (proto.has_annotation_map()) {
    ZETASQL_RETURN_IF_ERROR(params.type_factory->DeserializeAnnotationMap(
        proto.annotation_map(), &annotation_map));
  }
  return ResolvedColumn(static_cast<int>(proto.column_id()), table_name,
                        column_name, AnnotatedType(type, annotation_map));
}

std::string ResolvedColumnListToString(const ResolvedColumnList& columns) {
  if (columns.empty()) return "[]";
  const std::string& common_table_name = columns[0].table_name();
  // Use the regular format if we have only one column.
  bool use_common_table_name = (columns.size() > 1);
  for (int i = 1; i < columns.size(); ++i) {
    if (columns[i].table_name() != common_table_name) {
      use_common_table_name = false;
      break;
    }
  }

  std::string ret;
  if (use_common_table_name) {
    absl::StrAppend(&ret, common_table_name, ".[");
    for (const ResolvedColumn& column : columns) {
      if (&column != &columns[0]) ret += ", ";
      ret += column.ShortDebugString();
    }
  } else {
    ret += "[";
    for (const ResolvedColumn& column : columns) {
      if (&column != &columns[0]) ret += ", ";
      ret += column.DebugString();
    }
  }
  ret += "]";
  return ret;
}

}  // namespace zetasql
